sql: prohibit new data types from usage in arrays during upgrades

otan opened the pull request:
Fixed a bug where array types containing new types were not considered
during a version upgrade.

Resolves #53952 
Resolves #53014 

Release justification: low risk update to new functionality

Release note (bug fix): Fix a bug where we allowed new types to be used
in an array type during a version upgrade.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53961)
<!-- Reviewable:end -->

jordanlewis reviewed (approved):
Thanks!

otan committed:
sql: prohibit new data types from usage in arrays during upgrades

Fixed a bug where array types containing new types were not considered
during a version upgrade.

Release justification: low risk update to new functionality

Release note (bug fix): Fix a bug where we allowed new types to be used
in an array type during a version upgrade.
